includes/Modules/Sign_In_With_Google.php

```diff
@@ -36,12 +44,141 @@ final class Sign_In_With_Google extends Module implements Module_With_Assets, Mo
 	 */
 	const MODULE_SLUG = 'sign-in-with-google';
 
+	/**
+	 * Validate_Auth_Request instance.
+	 *
+	 * @since n.e.x.t
+	 * @var Validate_Auth_Request
+	 */
+	protected $validate_auth_request;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Context        $context        Plugin context.
+	 * @param Options        $options        Optional. Option API instance. Default is a new instance.
+	 * @param User_Options   $user_options   Optional. User Option API instance. Default is a new instance.
+	 * @param Authentication $authentication Optional. Authentication instance. Default is a new instance.
+	 * @param Assets         $assets         Optional. Assets API instance. Default is a new instance.
+	 */
+	public function __construct(
+		Context $context,
+		Options $options = null,
+		User_Options $user_options = null,
+		Authentication $authentication = null,
+		Assets $assets = null
+	) {
+		parent::__construct( $context, $options, $user_options, $authentication, $assets );
+
+		$this->validate_auth_request = new Validate_Auth_Request( $context );
+	}
+
 	/**
 	 * Registers functionality through WordPress hooks.
 	 *
 	 * @since 1.137.0
 	 */
 	public function register() {
+		add_action(
+			'init',
+			function () {
+				add_rewrite_rule( '^auth/google/?$', 'index.php?google_auth=1', 'top' );
+			}
+		);
+
+		add_filter(
+			'query_vars',
+			function ( $vars ) {
+				$vars[] = 'google_auth';
+				return $vars;
+			}
+		);
+
+		add_filter( 'wp_login_errors', array( $this, 'handle_google_auth_errors' ) );
+
+		add_action( 'template_redirect', array( $this, 'handle_google_auth' ) );
+	}
+
+	/**
+	 * Intercept the page request to process token ID
+	 * and complete Sign in with Google flow.
+	 *
+	 * @since n.e.x.t
+	 */
+	public function handle_google_auth() {
+		global $wp_query;
+
+		if ( ! isset( $wp_query->query_vars['google_auth'] ) ) {
+			return;
+		}
+
+		$client_id = $this->get_settings()->get()['clientID'];
+		if ( empty( $client_id ) ) {
+			wp_safe_redirect( add_query_arg( 'error', 'no_client_id', wp_login_url() ) );
+			exit;
+		}
+
+		$this->validate_auth_request->run_validations();
+		$error = $this->validate_auth_request->get_error();
+
+		if ( is_wp_error( $error ) ) {
+			wp_safe_redirect( add_query_arg( 'error', $error->get_error_code(), wp_login_url() ) );
+			exit;
+		}
+
+		$id_token = $this->context->input()->filter( INPUT_POST, 'credential' );
+		try {
+			$client  = new Google_Client( array( 'client_id' => $client_id ) );
+			$payload = $client->verifyIdToken( $id_token );
+
+			if ( empty( $payload ) ) {
+				wp_safe_redirect( add_query_arg( 'error', 'google_auth_invalid_request', wp_login_url() ) );
+				exit;
+			}
+
+			// @TODO implement further flow and redirect with $payload.
```

Comment:
@zutigrm just checking if this TODO needed to be addressed in this PR/issue or in a further issue.

Reply:
Thanks for checking, this will be extended in #9339

Reply:
Thank you!